packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils.ts

```diff
@@ -1,19 +1,60 @@
-// import { getPoolsWithTokens, parsePoolData } from '@balancer-labs/sor'; // TODO - upgrade to v2
 import { BigNumber } from '@0x/utils';
+// import { parsePoolData } from '@balancer-labs'; // TODO - upgrade to v2
 import { Pool } from '@balancer-labs/sor/dist/types';
-import { request } from 'graphql-request';
+import { gql, request } from 'graphql-request';
 
-import { BALANCER_MAX_POOLS_FETCHED, BALANCER_V2_SUBGRAPH_URL } from '../constants';
+import { DEFAULT_WARNING_LOGGER } from '../../../constants';
+import { LogFunction } from '../../../types';
+import { BALANCER_MAX_POOLS_FETCHED, BALANCER_TOP_POOLS_FETCHED, BALANCER_V2_SUBGRAPH_URL } from '../constants';
 
+import { parsePoolData } from './balancer_sor_v2';
 import { CacheValue, PoolsCache } from './pools_cache';
 
+// tslint:disable-next-line:custom-no-magic-numbers
+const ONE_DAY_MS = 24 * 60 * 60 * 1000;
+
+interface BalancerPoolResponse {
+    id: string;
+    swapFee: string;
+    tokens: Array<{ address: string; decimals: number; balance: string; weight: string; symbol: string }>;
+    tokensList: string[];
+    totalWeight: string;
+    totalShares: string;
+    amp: string | null;
+}
+
 export class BalancerV2PoolsCache extends PoolsCache {
+    private static _parseSubgraphPoolData(pool: any, takerToken: string, makerToken: string): Pool {
+        const tToken = pool.tokens.find((t: any) => t.address === takerToken);
+        const mToken = pool.tokens.find((t: any) => t.address === makerToken);
+        const swap = pool.swaps && pool.swaps[0];
+        const tokenAmountOut = swap ? swap.tokenAmountOut : undefined;
+        const tokenAmountIn = swap ? swap.tokenAmountIn : undefined;
+        const spotPrice =
+            tokenAmountOut && tokenAmountIn ? new BigNumber(tokenAmountOut).div(tokenAmountIn) : undefined; // TODO: xianny check
+
+        return {
+            id: pool.id,
+            balanceIn: new BigNumber(tToken.balance),
+            balanceOut: new BigNumber(mToken.balance),
+            weightIn: new BigNumber(tToken.weight),
+            weightOut: new BigNumber(mToken.weight),
+            swapFee: new BigNumber(pool.swapFee),
+            spotPrice,
+        };
+    }
+
     constructor(
         private readonly subgraphUrl: string = BALANCER_V2_SUBGRAPH_URL,
         private readonly maxPoolsFetched: number = BALANCER_MAX_POOLS_FETCHED,
+        private readonly _topPoolsFetched: number = BALANCER_TOP_POOLS_FETCHED,
+        private readonly _warningLogger: LogFunction = DEFAULT_WARNING_LOGGER,
         cache: { [key: string]: CacheValue } = {},
     ) {
         super(cache);
+        void this._loadTopPoolsAsync();
+        // Reload the top pools every 12 hours
+        setInterval(async () => void this._loadTopPoolsAsync(), ONE_DAY_MS / 2);
```

Comment:
why?

Reply:
If it wasn't discovered by the last deployment (on init), then the only way a pool can be discovered is through natural search. 

We want a method to discover decently liquid pools that users may not naturally search. Maybe a liquid BAL/USDT pool appears but no one ever searches BAL/USDT, they always search BAL/USDC.  We don't go searching on Multihop intermediary hops automatically.